File: frontend/library/filterSpec.js
// frontend/library/filterSpec.js
/** FilterSpec 클라이언트 평가기. backend/core/query/eval_inmem.py와 의미 일치(parity 픽스처로 강제). */
const DATE_FIELDS = new Set(['due_date', 'start_date', 'created_at', 'updated_at']);
const LEAF = { status: 'status', status_category: 'status_category', priority: 'priority',
  task_type: 'task_type', epic: 'epic_id', sprint: 'sprint_id', created_by: 'created_by',
  due_date: 'due_date', start_date: 'start_date', created_at: 'created_at', updated_at: 'updated_at' };

function addDays(iso, n) {
  const d = new Date(iso + 'T00:00:00Z');
  d.setUTCDate(d.getUTCDate() + n);
  return d.toISOString().slice(0, 10);
}
function resolve(value, ctx, isDate) {
  if (value === '$me') return ctx.userId;
  if (isDate && typeof value === 'string') {
    if (value === '$today') return ctx.today;
    // grammar는 filter_spec._DATE_TOKEN / eval_inmem._REL과 동기
    const m = /^\$today([+-]\d+)d$/.exec(value);
    if (m) return addDays(ctx.today, parseInt(m[1], 10));
  }
  return value;
}
function cmp(a, op, b) {
  if (a == null) return false;
  if (op === 'eq') return a === b;
  if (op === 'lt') return a < b;
  if (op === 'lte') return a <= b;
  if (op === 'gt') return a > b;
  if (op === 'gte') return a >= b;
  return false;
}
const strip = (s) => (s || '').replace(/<[^>]+>/g, ' ');

function cfText(raw) {
  // custom_fields raw → 비교용 텍스트(Py _cf_text와 1:1). bool 소문자, null은 비매칭.
  // 소수(1.0)는 JS가 int/float 구분이 없어 Py와 표기차가 남음 — v1 cf는 text/select 위주.
  if (raw == null) return null;
  if (typeof raw === 'boolean') return raw ? 'true' : 'false';
  return String(raw);
}

function evalCond(task, node, ctx) {
  const { field, op, value } = node;
  if (field === 'assignee') {
    if (op === 'is_empty') return (task.assignees || []).length === 0;
    const ids = (op === 'in' ? value : [value]).map((v) => (v === '$me' ? ctx.userId : v));
    return (task.assignees || []).some((a) => ids.includes(a.user_id));
  }
  if (field === 'label') {
    if (op === 'is_empty') return (task.labels || []).length === 0;
    const set = new Set(value);
    return (task.labels || []).some((l) => set.has(l.label_id));
  }
  if (field === 'has_subtasks') {
    const has = (task.subtaskCount || (task.subtasks || []).length) > 0;
    return value ? has : !has;
  }
  if (field === 'is_top_level') {
    const top = task.parent_task_id == null;
    return value ? top : !top;
  }
  if (field === 'text') {
    const hay = `${task.title || ''} ${strip(task.description)}`.toLowerCase();
    return op === 'eq' ? hay.trim() === String(value).toLowerCase() : hay.includes(String(value).toLowerCase());
  }
  if (field.startsWith('cf:')) {
    const raw = cfText((task.custom_fields || {})[field.slice(3)]);
    if (op === 'is_empty') return raw == null;
    if (raw == null) return false;
    if (op === 'contains') return raw.toLowerCase().includes(String(value).toLowerCase());
    if (op === 'in') return value.map(String).includes(raw);
    // eq (v1: cf 비교 op 없음)
    return raw === String(value);
  }
  const col = LEAF[field];
  const isDate = DATE_FIELDS.has(field);
  const lv = task[col];
  if (op === 'is_empty') return lv == null;
  if (op === 'in') return value.map((v) => resolve(v, ctx, isDate)).includes(lv);
  if (op === 'between') return lv != null && lv >= resolve(value[0], ctx, isDate) && lv <= resolve(value[1], ctx, isDate);
  return cmp(lv, op, resolve(value, ctx, isDate));
}

export function evaluate(task, node, ctx = {}) {
  if (!node) return true;
  let res;
  if (node.type === 'group') {
    const ch = node.children || [];
    if (ch.length === 0) res = true;
    else if (node.op === 'OR') res = ch.some((x) => evaluate(task, x, ctx));
    else res = ch.every((x) => evaluate(task, x, ctx));
  } else {
    res = evalCond(task, node, ctx);
  }
  return node.negate ? !res : res;
}

File: backend/core/query/filter_spec.py
"""FilterSpec 구조 검증(DB 무관). SQL 빌더·평가기가 공유하는 화이트리스트 계약."""
import datetime as _dt
import re

MAX_DEPTH = 5
MAX_NODES = 50

FIELD_SPECS = {
    "status": {"type": "enum", "ops": {"eq", "in", "is_empty"}},
    "status_category": {"type": "enum", "ops": {"eq", "in"}},
    "priority": {"type": "enum", "ops": {"eq", "in"}},
    "task_type": {"type": "enum", "ops": {"eq", "in"}},
    "label": {"type": "id", "ops": {"in", "is_empty"}},
    "epic": {"type": "id", "ops": {"eq", "in", "is_empty"}},
    "sprint": {"type": "id", "ops": {"eq", "in", "is_empty"}},
    "assignee": {"type": "id", "ops": {"eq", "in", "is_empty"}},
    "created_by": {"type": "id", "ops": {"eq", "in"}},
    "due_date": {"type": "date", "ops": {"eq", "lt", "lte", "gt", "gte", "between", "is_empty"}},
    "start_date": {"type": "date", "ops": {"eq", "lt", "lte", "gt", "gte", "between", "is_empty"}},
    "created_at": {"type": "datetime", "ops": {"lt", "lte", "gt", "gte", "between"}},
    "updated_at": {"type": "datetime", "ops": {"lt", "lte", "gt", "gte", "between"}},
    "text": {"type": "text", "ops": {"contains", "eq"}},
    "has_subtasks": {"type": "bool", "ops": {"eq"}},
    "is_top_level": {"type": "bool", "ops": {"eq"}},
}
_CUSTOM_OPS = {"eq", "in", "contains", "is_empty"}  # v1: 비교/between 제외 — JSONB ->> 는 text라 숫자 비교 오답("10"<"2") 방지. 숫자/날짜 cf 비교는 후속(field_type cast)


class FilterError(ValueError):
    pass


def is_custom_field(field):
    return isinstance(field, str) and field.startswith("cf:") and field[3:].isdigit()


def field_type(field):
    if is_custom_field(field):
        return "custom"
    fs = FIELD_SPECS.get(field)
    return fs["type"] if fs else None


def _ops_for(field):
    if is_custom_field(field):
        return _CUSTOM_OPS
    fs = FIELD_SPECS.get(field)
    return fs["ops"] if fs else None


def validate_filter(spec):
    if spec is None:
        return None
    if _walk(spec, 1) > MAX_NODES:
        raise FilterError(f"too many nodes (>{MAX_NODES})")
    return None


def _walk(node, depth):
    if depth > MAX_DEPTH:
        raise FilterError(f"filter too deep (>{MAX_DEPTH})")
    if not isinstance(node, dict):
        raise FilterError("node must be an object")
    if "negate" in node and not isinstance(node["negate"], bool):
        raise FilterError("negate must be a boolean")
    if node.get("type") == "group":
        if node.get("op") not in ("AND", "OR"):
            raise FilterError("group op must be AND or OR")
        children = node.get("children") or []
        if not isinstance(children, list):
            raise FilterError("children must be a list")
        return 1 + sum(_walk(c, depth + 1) for c in children)
    if node.get("type") == "cond":
        ops = _ops_for(node.get("field"))
        if ops is None:
            raise FilterError(f"unknown field: {node.get('field')}")
        if node.get("op") not in ops:
            raise FilterError(f"op {node.get('op')} not allowed for {node.get('field')}")
        _check_value(node.get("op"), node.get("value"))
        _check_typed(node.get("field"), node.get("op"), node.get("value"))
        return 1
    raise FilterError("node type must be 'group' or 'cond'")


def _check_value(op, value):
    if op == "is_empty" and value is not None:
        raise FilterError("'is_empty' takes no value")  # 잘못된 페이로드가 계약을 통과하지 않도록
    if op == "in" and not (isinstance(value, list) and len(value) > 0):
        raise FilterError("'in' requires a non-empty list")  # 빈 리스트 → SQL IN () 500 차단
    if op == "between" and not (isinstance(value, list) and len(value) == 2):
        raise FilterError("'between' requires [from, to]")


# $today, $today+7d, $today-3d. 이 grammar는 eval_inmem._REL과 동기 유지(거기선 정수 오프셋만 캡처).
_DATE_TOKEN = re.compile(r"^\$today([+-]\d+d)?$")


def _is_date_value(v):
    # date/datetime 공통: ISO 날짜(YYYY-MM-DD) 또는 $today 토큰만. datetime도 v1은 날짜 단위 비교만
    # 지원(타임스탬프 입력 미지원 — 문자열 비교가 Py/JS/SQL에서 동일 동작).
    if not isinstance(v, str):
        return False
    if _DATE_TOKEN.match(v):
        return True
    try:
        _dt.date.fromisoformat(v)
        return True
    except ValueError:
        return False


def _check_typed(field, op, value):
    """필드 타입별 값 검증 — 잘못된 값이 SQL까지 내려가 500 나는 것을 INVALID_FILTER로 차단."""
    if op == "is_empty":
        return
    ftype = field_type(field)
    vals = value if op in ("in", "between") else [value]
    for v in vals:
        if ftype in ("enum", "text", "custom"):
            if not isinstance(v, str):
                raise FilterError(f"{field}: value must be a string")
        elif ftype == "id":
            allow_me = field in ("assignee", "created_by")
            if not (isinstance(v, int) and not isinstance(v, bool)) and not (allow_me and v == "$me"):
                raise FilterError(f"{field}: value must be an integer id" + (" or '$me'" if allow_me else ""))
        elif ftype in ("date", "datetime"):
            if not _is_date_value(v):  # $me 등 비날짜 값 거부 — $me→user_id로 치환돼 잘못된 비교가 SQL로 새는 것 차단
                raise FilterError(f"{field}: value must be ISO date (YYYY-MM-DD) or a $today token")
        elif ftype == "bool":
            if not isinstance(v, bool):
                raise FilterError(f"{field}: value must be boolean true/false")
